### ISSUE TITLE
RFC: Error if requested secret does not exist

### ISSUE THREAD

**Issue by bradrydzewski** [harness/harness#2423]
This is a proposal to change the current behavior and error a step if the requested secret does not exist. To illustrate, we will assume a repository has the following secrets defined:

```shell
$ drone secret ls
DOCKER_USERNAME
DOCKER_PASSWORD
```

And the repository uses the below yaml, which would result in a pipeline failure and error message because the DOCKER_EMAIL secret does not exist.

```diff
pipeline:
  publish:
    image: plugins/docker
    secrets:
      - DOCKER_USERNAME 
      - DOCKER_PASSWORD
-     - DOCKER_EMAIL
```

The only gotcha is that a secret may be defined, but may only be accessible to specific events. For example, the above yaml could fail if the event type is pull request, but the secret is limited to push events. The yaml would have to be adjusted to limit step execution in order to avoid failure:

```diff
pipeline:
  publish:
    image: plugins/docker
    secrets:
      - DOCKER_USERNAME 
      - DOCKER_PASSWORD
+   when:
+     event: push
```

This behavior change could break some configurations, although it is probably unlikely. I wanted to check with everyone before I proceed. Let me know if this would cause an issue for your organization, or if there are any major blockers that I am not considering.

**Comment by bradrydzewski** [comment]
cc @gtaylor @tonglil @metalmatze @tboerger @jimsheldon

**Comment by tboerger** [comment]
Sounds like a useful improvement to me :+1:

**Comment by frebib** [comment]
Many times we have had failures due to missing secrets but not known why. This sounds like good idea

**Comment by gtaylor** [comment]
We're definitely pro-failing-noisily-with-helpful-error!

**Comment by bradrydzewski** [comment]
ok great, lets plan on including this in 0.9 ... probably in rc.2

**Comment by tonglil** [comment]
Yes, erroring out (with an appropriate message) for a missing secret is 👍 

The message should say something like `secret "XYZ" not available for "abc" step`.

**Comment by skolodyazhnyy** [comment]
Wrong events for secrets are one of very common drone misconfigurations for us. I'm willing to contribute and this issue seems like something easy to pick up. Would it make sense for me to try to implement it? If so, do you have any advice on how to do it best?

**Comment by bradrydzewski** [comment]
I am still working out the design. This would be a breaking change for a number of projects, therefore, I think we need an alternate approach.

Specifically there are teams that configure secrets and pipeline steps knowing the secret may or may not be available, and then they check for that secret at runtime. We do not want to break these types of configurations (contrived example below)

```
pipeline:
  build:
    image: golang
    secrets: [ docker_password ]
    commands:
      - [ -z $DOCKER_PASSWORD ] && ./run.sh
```

One option is to introduce a concept of warnings, allowing the build to proceed even though the secret is not available, and then display useful warning messages in the user interface. Another option is to add some sort of "strict mode" parameter that instructs drone to error when the secret is not available. I'm not a huge fan of either option, and would like to collect some more ideas before we settle on an approach.

**Comment by jmccann** [comment]
Maybe a `when` for if a secret exists?

above then becomes:

```
pipeline:
  test:
    image: golang
    secrets: [ private_key ]
    commands:
      - go get
      - go test

  build:
    when:
      secrets:
        include: [ private_key ]
    commands:
      - ./build_and_sign.sh
```

**Comment by bradrydzewski** [comment]
drone 1.0 improved how secrets are defined and used by making the syntax more explicit. The syntax in 0.8 was confusing which caused a support burden. Comparing the two versions:

before:

```text
pipeline:
  publish:
    image: plugins/docker
    secrets: [ DOCKER_USERNAME, DOCKER_PASSORD ]
    repo: octocat/hello-world
    tags: latest
```
after:

```text
- name: publish
  image: plugins/docker
  settings:
    username:
      from_secret: docker_username
    password:
      from_secret: docker_password
    repo: octocat/hello-world
    tags: latest
```

Anecdotally this seems to have completely eliminated a class of configuration errors where the end-user incorrectly named the secrets in 0.8. This may obsolete this particular request, although I think we need to observe usage over the next few months before we make a decision.

If we decide to move forward and make secrets required, we could support a basic syntax and expanded syntax allowing users to override the default behavior and make secrets optional:

basic syntax

```text
- name: publish
  image: plugins/docker
  settings:
    username:
      from_secret: docker_username
```

expanded syntax:

```text
- name: publish
  image: plugins/docker
  settings:
    username:
      from_secret:
        name: docker_username
        optional: true
```

**Comment by foosinn** [comment]
Can you give some implementation hints for this?

I assume all github.com/drone/drone-yaml/yaml/compiler/transform should change from `func(*engine.Spec)` to `func(*engine.Spec) error`?
